File: fs.py
import os
#from struct import pack, unpack
import struct
pack, unpack = struct.pack, struct.unpack
#import argparse
import sys
#cwd = os.path.dirname(__file__)


def get_str(inbytes, index=0):
    byte = inbytes[index]
    data = b""
    while byte != 0:
        index += 1
        data += bytes([byte])
        byte = inbytes[index]

    return data.decode("utf-8")


fmt = "=I"


def writeimg(filedir="testfiles", imgfile="pyfs.img"):
    cwd = os.getcwd()
    files = []
    os.chdir(filedir)
    for x in os.listdir('.'):
        try:
            with open(x, "rb") as f:
                read: bytes = f.read()
                files.append([bytes([ord(y) for y in x]) + b"\0",
                 pack(fmt, len(read)), read])
        except Exception as e:
            #print('PermissionError:',e)
            pass
    #print(files)

    binfiles = []
    for x in files:
        binfiles.append(b"".join(x))

    #print(binfiles)
    data = b"".join(binfiles)
    #print(data)
    os.chdir(cwd)
    with open(imgfile, "wb") as f:
        f.write(data)


def readimg(filedir="testread", imgfile="pyfs.img"):
    cwd = os.getcwd()
    files = []
    with open(imgfile, "rb") as f:
        read: bytes = f.read()
    #print(read)
    index: int = 0
    while True:
        try:
            filename: str = get_str(read[index:])
            # assert filename != ''
            index += len(filename) + 1
            length: int = unpack(fmt, read[index : index + 4])[0]
            index += 4
            data: bytes = read[index : index + length]
            index += length
            files.append([filename, length, data])
            # raise Exception()

        except Exception as e:
            # print(e)
            break
    #print(files)
    os.chdir(filedir)
    for x in files:
        with open(x[0], "wb") as f:
            f.write(x[2])
    os.chdir(cwd)

#parser = argparse.ArgumentParser(description="Read from and write to a pyfs image file")
#subparsers = parser.add_subparsers()
#read = subparsers.add_parser('read', help='read from an image file')
#read.add_argument('img',default='pyfs.img',help="image file to read from")
#read.add_argument('dir',default='testread',help="directory to write to")
#read.set_defaults(func=readimg)

#write = subparsers.add_parser('write', help='write to an image file')
#write.add_argument('img',default='pyfs.img',help="image file to write to")
#write.add_argument('dir',default='testfiles',help="directory to read from")
#write.set_defaults(func=writeimg)

#args = parser.parse_args()

#print(not not args.read,not not args.write)
#args.func(args.dir,args.img)
args=sys.argv[1:]
helpmsg = '''usage: pyfs.py [-h] {read,write} img dir

Read from and write to a pyfs image file

positional arguments:
  {read,write}
    read        read from an image file
    write       write to an image file
    img         image file to read/write from
    dir         directory to write/read to
options:
  -h, --help    show this help message and exit'''
if len(args) > 2:
    if args[0].lower() == 'read':
        print('read')
        readimg(args[2],args[1])
    elif args[0].lower() == 'write':
        writeimg(args[2],args[1])
    else:
        print(helpmsg)
else:
    print(helpmsg)
#print(args,len(args) > 2)

File: fs.cpp
#include "builtin.hpp"
#include "os/path.hpp"
#include "os/__init__.hpp"
#include "stat.hpp"
#include "sys.hpp"
#include "struct.hpp"
#include "fs.hpp"

namespace __fs__ {

str *const_0, *const_1, *const_2, *const_3, *const_4, *const_5, *const_6, *const_7, *const_8, *const_9;




list<str *> *args;
str *__name__, *fmt, *helpmsg;
void *f;

str * default_0;
str * default_1;
str * default_2;
str * default_3;


static inline list<__ss_int> *list_comp_0(pyseq<pyseq<pyobj *> *> *x);

static inline list<__ss_int> *list_comp_0(pyseq<pyseq<pyobj *> *> *x) {
    pyseq<pyobj *> *y;
    pyseq<pyseq<pyobj *> *> *__9;
    __iter<pyseq<pyobj *> *> *__10;
    __ss_int __11;
    pyseq<pyseq<__ss_int> *>::for_in_loop __12;

    list<__ss_int> *__ss_result = new list<__ss_int>();

    FOR_IN(y,x,9,11,12)
        __ss_result->append(ord(y));
    END_FOR

    return __ss_result;
}

void *writeimg(str *filedir, str *imgfile) {
    str *cwd;
    list<list<bytes *> *> *__13, *files;
    pyseq<pyseq<pyobj *> *> *x;
    void *__7, *__8, *read;
    list<bytes *> *binfiles;
    bytes *data;
    list<str *> *__3;
    __iter<str *> *__4;
    __ss_int __15, __5;
    list<str *>::for_in_loop __6;
    file_binary *f;
    __iter<list<bytes *> *> *__14;
    list<list<bytes *> *>::for_in_loop __16;

    cwd = __os__::getcwd();
    files = (new list<list<bytes *> *>());
    __os__::chdir(filedir);

    FOR_IN(x,__os__::listdir(const_0),3,5,6)
        try {
            WITH_VAR(open_binary(x, const_1),f,0)
                files->append((new list<bytes *>(3,(__bytes(list_comp_0(x)))->__add__(new bytes("\000", 1)),pack(1, __fs__::fmt, len(read)),read)));
            END_WITH
        } catch (Exception *e) {
        }
    END_FOR

    binfiles = (new list<bytes *>());

    FOR_IN(x,files,13,15,16)
        binfiles->append((new bytes(""))->join(x));
    END_FOR

    data = (new bytes(""))->join(binfiles);
    __os__::chdir(cwd);
    WITH_VAR(open_binary(imgfile, const_2),f,1)
        f->write(data);
    END_WITH
    return NULL;
}

void *readimg(str *filedir, str *imgfile) {
    str *cwd;
    list<list<void *> *> *__17, *files;
    void *data, *filename, *index, *length, *read;
    list<void *> *x;
    file_binary *f;
    __iter<list<void *> *> *__18;
    __ss_int __19;
    list<list<void *> *>::for_in_loop __20;

    cwd = __os__::getcwd();
    files = (new list<list<void *> *>());
    WITH_VAR(open_binary(imgfile, const_1),f,2)
    END_WITH

    while (True) {
        try {
            index = (index)->__iadd__((len(filename))->__add__(__ss_int(1)));
            index = (index)->__iadd__(__ss_int(4));
            index = (index)->__iadd__(length);
            files->append((new list<void *>(3,filename,length,data)));
        } catch (Exception *e) {
            break;
        }
    }
    __os__::chdir(filedir);

    FOR_IN(x,files,17,19,20)
        WITH_VAR(open_binary(x->__getfast__(__ss_int(0)), const_2),f,3)
            f->write(x->__getfast__(__ss_int(2)));
        END_WITH
    END_FOR

    __os__::chdir(cwd);
    return NULL;
}

void __init() {
    const_0 = __char_cache[46];
    const_1 = new str("rb");
    const_2 = new str("wb");
    const_3 = new str("=I");
    const_4 = new str("testfiles");
    const_5 = new str("pyfs.img");
    const_6 = new str("testread");
    const_7 = new str("usage: pyfs.py [-h] {read,write} img dir\n\nRead from and write to a pyfs image file\n\npositional arguments:\n  {read,write}\n    read        read from an image file\n    write       write to an image file\n    img         image file to read/write from\n    dir         directory to write/read to\noptions:\n  -h, --help    show this help message and exit");
    const_8 = new str("read");
    const_9 = new str("write");

    __name__ = new str("__main__");

    fmt = const_3;
    default_0 = const_4;
    default_1 = const_5;
    default_2 = const_6;
    default_3 = const_5;
    args = (__sys__::argv)->__slice__(__ss_int(1), __ss_int(1), __ss_int(0), __ss_int(0));
    helpmsg = const_7;
    if ((len(__fs__::args)>__ss_int(2))) {
        if (__eq((__fs__::args->__getfast__(__ss_int(0)))->lower(), const_8)) {
            print(1, NULL, NULL, NULL, const_8);
            readimg(__fs__::args->__getfast__(__ss_int(2)), __fs__::args->__getfast__(__ss_int(1)));
        }
        else {
            if (__eq((__fs__::args->__getfast__(__ss_int(0)))->lower(), const_9)) {
                writeimg(__fs__::args->__getfast__(__ss_int(2)), __fs__::args->__getfast__(__ss_int(1)));
            }
            else {
                print(1, NULL, NULL, NULL, __fs__::helpmsg);
            }
        }
    }
    else {
        print(1, NULL, NULL, NULL, __fs__::helpmsg);
    }
}

} // module namespace

int main(int __ss_argc, char **__ss_argv) {
    __shedskin__::__init();
    __stat__::__init();
    __os__::__path__::__init();
    __os__::__init();
    __struct__::__init();
    __sys__::__init(__ss_argc, __ss_argv);
    __shedskin__::__start(__fs__::__init);
}
